### ISSUE TITLE
XCUITestDriver returns StaleElementReferenceError: Element does not exist in cache

### ISSUE THREAD

**Issue by ilendemli** [appium/appium#18839]
### Do I have the most recent component updates?

- [X] I use the most recent available driver/plugin and server versions

### Is the component officially supported by the Appium team?

- [X] I have verified the component repository is present under the Appium organization in GitHub

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

For some reason, after quickly switching back and forth between the NATIVE_APP and WEBVIEW context (doing getContexts before switching) in our test cases, locating elements fail because of StaleElementReferenceError. `getPageSource` fails with `Method threw 'org.openqa.selenium.WebDriverException' exception.`
```
[debug] [ios-device] Received message from Web Inspector:
[debug] [ios-device] {
[debug] [ios-device]   "__argument": {
[debug] [ios-device]     "WIRApplicationIdentifierKey": "PID:33769",
[debug] [ios-device]     "WIRMessageDataTypeKey": "WIRMessageDataTypeFull",
[debug] [ios-device]     "WIRDestinationKey": "d5038e8c-c470-4150-989e-ccb31941c4fd",
[debug] [ios-device]     "WIRMessageDataKey": "{\"method\":\"Target.dispatchMessageFromTarget\",\"params\":{\"targetId\":\"page-27\",\"message\":\"{\\\"result\\\":{\\\"result\\\":{\\\"type\\\":\\\"object\\\",\\\"objectId\\\":\\\"{\\\\\\\"injectedScriptId\\\\\\\":1,\\\\\\\"id\\\\\\\":2}\\\",\\\"className\\\":\\\"k\\\",\\\"description\\\":\\\"StaleElementReferenceError: Element does not exist in cache\\\"},\\\"wasThrown\\\":true},\\\"id\\\":298}\"}}"
[debug] [ios-device]   },
[debug] [ios-device]   "__selector": "_rpc_applicationSentData:"
[debug] [ios-device] }
```
The webpage is shown in a hybrid app. It is inspectable through Safari.

### Expected Behavior

Getting Element after switching context to WEBVIEW should not fail.

### Minimal Reproducible Example

I have no example to provide at the moment but will try to create one to reproduce the issue.

### Environment

- Operating system: MacOS & iOS
- Appium server version (output of `appium --version`): 2023.5.2
- Appium driver(s) and their version(s): WDA 5.5.1
- Appium plugin(s) and their version(s):
- Node.js version (output of `node --version`): 20.2.0
- `npm` version (output of `npm --version`): 9.6.6
- Last component(s) version which did _not_ exhibit the problem:
- Platform and version under test: iOS 16.5
- Real device or emulator/simulator: Real device

**Comment by mykola-mokhnach** [comment]
Maybe the internal elements cache is refreshed every time the web view context is switched. Try to locate elements instead of using cached ones as a workaround

**Comment by mykola-mokhnach** [comment]
Also, it is not possible to help more without the full server log

**Comment by ilendemli** [comment]
We are already using locators as far as i know. I will try to provide a log but have to redact a lot.

**Comment by mykola-mokhnach** [comment]
> We are already using locators as far as i know.

sorry, I don't get that. Locators are required anyway. What I am talking about is instead of using something like:

```
element = driver.find_element(locator)
text = element.text
# do some context switching
element.click()
```

change it to

```
# do some context switching
text = driver.find_element(locator).text
# do some context switching
driver.find_element(locator).click()
```

**Comment by mykola-mokhnach** [comment]
Closed because of no response

**Comment by ilendemli** [comment]
Hey, sorry I haven't had any time looking into this. In the meantime I downgraded nodejs to 16.20.1 and nom to 8.19.4 because I had some issues connecting to my phone. We still have the problem so please reopen this. Here is the server log with our informations redacted: https://gist.github.com/ilendemli/30afdc6719755648466c9036d5dcb0cb. If you need any assistant regarding this please contact me and I can share more information privately.

We are basically doing the same test twice in one run, it hangs during the second run, where there is a new WebView context.

**Comment by mykola-mokhnach** [comment]
The log above shows that Appium 1 is used, which we don't support. Please upgrade to Appium 2 first. Also, have you applied the above workaround in your client code?

**Comment by ilendemli** [comment]
I will try to get this running with Appium 2 again soon and provide you a log. Afaik I had the same issue with it back when I managed to run this. We also tried the workaround, but in our case we first switch the context when try to find elements using the locator, so it's basically:
```
# do some context switching
driver.find_element(locator).click()
```

**Comment by ilendemli** [comment]
@mykola-mokhnach here is the server log from appium 2 running the same test: https://gist.github.com/ilendemli/c89c0ba717e0fe5284c0daa9c6d69702

**Comment by jlipps** [comment]
The element you're trying to get the text of is an alert? Alerts are typically native elements (not webview ones), so I would work with the alert in the native context.

**Comment by ilendemli** [comment]
@jlipps no, line #6433 i get current contexts, 6440 i set the context, 6481 i try to find the element using css selector in webcontext, fails, but should not, ignore the infobubble alert stuff for the moment

**Comment by jlipps** [comment]
OK, the error appears to be happening when Appium is checking whether an alert is obscuring the webview. @mykola-mokhnach can you tell whether this error is coming from WDA or the driver?

**Comment by mykola-mokhnach** [comment]
This error is coming from the remote debugger or the atom itself. I cannot say for sure.

**Comment by jlipps** [comment]
@mykola-mokhnach are you sure?

```
[debug] [XCUITestDriver@0aa1 (57002fb2)] Proxying [GET /alert/text] to [GET http://127.0.0.1:8100/session/D4D39557-3ABC-4DCE-9656-687C3C9741BC/alert/text] with no body
```

This looks like it's being proxied to WDA

**Comment by mykola-mokhnach** [comment]
Yes, I am 100% sure.

> [debug] [XCUITestDriver@0aa1 (57002fb2)] Encountered internal error running command: Error: Remote debugger error with code 'undefined': StaleElementReferenceError: Element does not exist in cache

This error might only come from the remote debugger

**Comment by ilendemli** [comment]
I would love to help but I am not sure how to setup the IDE to run and debug appium and its node_modules. My breakpoint that I set in the module do not trigger.

**Comment by scottpage1317** [comment]
Is the element you're trying to locate deeply nested? The only time I see the StaleElementReferenceError on iOS is because of [this](https://github.com/appium/appium/issues/14825) issue

**Comment by ilendemli** [comment]
@scottpage1317 Not sure that is the reason, as it works for the first run, but does not for the second. To get an overview: One Test case, where the same steps are run a second time. Presses a button natively to open a webview, navigates through a form setting the fields, submits, does a purchase, all while switching between native and web context for each input, then dismisses the WebView. And then repeats the same steps. Goes trough up to until this issue happens.

**Comment by ilendemli** [comment]
I can reproduce the alert/text log reliable, but I am not sure if that related to the StaleElementReferenceError:
<img width="1793" alt="image" src="https://github.com/appium/appium/assets/2271750/5a110d1a-adf8-428e-bd80-f9db6f92ba3d">

it basically happens when you do not switch to native context before closing the webview
Example project: [Archive.zip](https://github.com/appium/appium/files/12043976/Archive.zip)
Log: https://gist.github.com/ilendemli/d66b866407dac72ee269faf5e84f16de

Steps to reproduce: Run app with appium inspector, open webview, switch context to webview, close webview, see server log

**Comment by ilendemli** [comment]
I managed to get StaleElementReferenceError with the example project provided above. Create a test case where you
1. press the button to open the webview
2. switch context to webview
3. get frame lement
4. switch to frame element
5. get html elements text
6. switch back to native context
7. close webview

repeat steps 1-7 in a loop

**Comment by jlipps** [comment]
thanks @ilendemli a repro is useful. the investigation/solution might take some time as there is a lot of updating work that needs to be done on the remote debugger code and the selenium atoms.

**Comment by mochi08** [comment]
Can you try setting your `snapshotMaxDepth` to 62 and see if it works?  For our app we were using 100 before and it was working fine, but recently it's giving us the "stale element reference" error.  What I discovered is that setting the `snapshotMaxDepth` value to a number larger than 62 will result in this error.

**Comment by ilendemli** [comment]
Sorry for the late response, we were in the process to move everything to Appium 2. We now have a Selenium Grid 4 running with Appium 2 and we still have the issue. `snapshotMaxDepth` does not help. I think the main issue is the WebView process being killed while still being in the WebView Context. Appium Inspector hangs in there as well. In our case it also fails to switch to the new process' context. Feel free to contact me if someone wants to have a private talk where I demonstrate the issue.

**Comment by rolandmoldovan** [comment]
> Can you try setting your `snapshotMaxDepth` to 62 and see if it works? For our app we were using 100 before and it was working fine, but recently it's giving us the "stale element reference" error. What I discovered is that setting the `snapshotMaxDepth` value to a number larger than 62 will result in this error.

Did you found a solution to the issue that you have described? I'm having the same problem and never found a solution/workaround.

**Comment by tamaniniandre** [comment]
Same issue here!

**Comment by SMasekoSTS** [comment]
Did anyone ever get a resolution for this? I am also experiencing this.

**Comment by mykola-mokhnach** [comment]
Per the thread, the element cache resets because the atoms script is re-injected fresh (a new
top-level JS scope, hence a brand-new cache object in `core/inject.ts`) on each context
switch. While the native context exists in WDA for the whole session lifecycle.
This is how the system was always designed and I don't think there is an easy way to change this behaviour. 
The only viable solution for now: never rely on cached web elements after switching web contexts. Found elements are only valid
while the context is active.
